x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/InferenceUtils.java

```diff
@@ -0,0 +1,133 @@
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public final class InferenceUtils {
+
+    private InferenceUtils() {}
+
+    public static void setInferenceResults(
+        LogicalPlan plan,
+        Client client,
+        InferenceContext inferenceContext,
+        ActionListener<Result> listener,
+        BiConsumer<LogicalPlan, ActionListener<Result>> callback
+    ) {
+        Set<SemanticQuery> semanticQueries = semanticQueries(plan);
+
+        if (semanticQueries.isEmpty()) {
+            callback.accept(plan, listener);
+            return;
+        }
+
+        GroupedActionListener<InferenceAction.Response> actionListener = new GroupedActionListener<>(
+            semanticQueries.size(),
+            new ActionListener<Collection<InferenceAction.Response>>() {
+                @Override
+                public void onResponse(Collection<InferenceAction.Response> ignored) {
+                    try {
+                        callback.accept(plan, listener);
+                    } catch (Exception e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            }
+        );
+        for (SemanticQuery semanticQuery : semanticQueries) {
+            InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                TaskType.ANY,
+                inferenceContext.semanticTextInferenceId(semanticQuery.fieldName()),
+                null,
+                List.of(semanticQuery.queryString()),
+                Map.of(),
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API,
+                false
+            );
+
+            executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                InferenceAction.INSTANCE,
+                inferenceRequest,
+                actionListener.delegateFailureAndWrap((next, inferenceResponse) -> {
+                    InferenceResults inferenceResults = SemanticTextInferenceUtils.validateAndConvertInferenceResults(
+                        inferenceResponse.getResults(),
+                        semanticQuery.fieldName()
+                    );
+                    setInferenceResult(plan, semanticQuery.fieldName(), semanticQuery.queryString(), inferenceResults);
+                    next.onResponse(inferenceResponse);
+                })
+            );
+        }
+    }
+
+    private record SemanticQuery(String fieldName, String queryString) {};
+
+    private static Set<SemanticQuery> semanticQueries(LogicalPlan analyzedPlan) {
+        Set<SemanticQuery> result = new HashSet<>();
+
+        analyzedPlan.forEachDown(plan -> {
+            if (plan instanceof Filter) {
+                ((Filter) plan).condition().forEachDown(expression -> {
```

Comment:
Why not execute it on Match instances with forEachDown(Match.class, match ->) ? Do you want to limit the inference action to match functions contained on a filter?

Reply:
That's a great point - will address!